Iquerejeta/pruning

iquerejeta opened the pull request:
Pruning of v0.3.0 of halo2: 

* Implements Plonk generically parametrised by a base field and a polynomial commitment scheme. 
* Uses the multi-open techniques of Halo2 (presented in their book) in the KZG instantiation
* Rewrites trait/interface of the PCS
* Removes IPA implementation
* Removes implementation of SHPLONK
* Removes implementation of shuffle argument

Note: I've rename the structure `Assigned` to `Rational`, as IMO it had nothing to do with assigned values. If you think differently, let me know. I can bring that back.

iquerejeta committed:
Remove book and halo2. Move everything to parent repo.

iquerejeta committed:
Remove Shuffle

iquerejeta committed:
Remove IPA

iquerejeta committed:
Remove SHPLONK

iquerejeta committed:
Remove compressed selectors

iquerejeta committed:
Temporarily remove examples and benches

iquerejeta committed:
Working version of new polynomial commitment scheme interface.

iquerejeta committed:
Plonk working parametrised only by F and PCS.

iquerejeta committed:
Bring examples and benches back

miguel-ambrona commented on src/arithmetic.rs:
```diff
@@ -507,6 +509,32 @@ pub(crate) fn powers<F: Field>(base: F) -> impl Iterator<Item = F> {
+    /// Add another multiexp into this one
+    fn add_msm(&mut self, other: &Self)
+    where
+        Self: Sized;
```
Can we require sized above instead?

miguel-ambrona commented on src/plonk/vanishing/prover.rs:
```diff
@@ -0,0 +1,175 @@
+        self,
+        x: F,
+        _xn: F,
+        _domain: &EvaluationDomain<F>,
```
Remove underscore.

miguel-ambrona commented on src/plonk/vanishing/prover.rs:
```diff
@@ -0,0 +1,175 @@
+            // .fold(domain.empty_coeff(), |acc, eval| acc * xn + eval);
+            .try_for_each(|p| {
+                let random_eval = eval_polynomial(p, x);
+                transcript.write(&random_eval)
```
Let's find a different name for this variable, other than `random_eval`. How about `h_piece_eval`?

miguel-ambrona commented on src/plonk/verifier.rs:
```diff
@@ -0,0 +1,310 @@
+        // Compute the expected value of h(x)
+        let expressions = advice_evals
+            .iter()
+            .zip(instance_evals.iter())
```
Remove `instance_evals`.

iquerejeta replied:
I think we need to keep them here.

iquerejeta committed:
Update to halo2curves 0.7.0

iquerejeta committed:
Apply tachyon optimizations

https://github.com/privacy-scaling-explorations/halo2/pull/342

iquerejeta committed:
Cost model update

iquerejeta committed:
* Implement PartialEq, Eq, Hash for Cell and AssignedCell

* Add table, compressed and normal rows count.

* Add rows and table rows to cost model.

* Ignore unassigned cells if they are multiplied by zero

* Some format values are written as

"Scalar(0x..)"

The hotfix was to change the stripping rules, but this is probably
an incorrect implementation of certain traits for one of the curves.

iquerejeta committed:
Review comments

iquerejeta committed:
Just moving things around!

Used the same style as we have in midnight-circuits.

Modules are defined with a `mod.rs` file inside the folder, instead of with a file at root level with the same name.

I've also created a module for utils and one for the transcript.

davidnevadoc commented on src/poly/commitment.rs:
```diff
@@ -0,0 +1,61 @@
+    /// This commits to a polynomial using its evaluations over the $2^k$ size
+    /// evaluation domain. The commitment will be blinded by the blinding factor
+    /// `r`.
+    fn commit_lagrange(&self, poly: &Polynomial<F, LagrangeCoeff>) -> CS::Commitment;
```
There may be a problem here inhereted from the previous version of the commitment scheme parameters.
If `Params` is implemented for the verifier parameters, will it always be possible to commit with such params?
For example, the verifier parameters for KZG should be only [s]_2. That's all that is needed to check an opening. However, with only this we cannot compute commitments.

I guess the underlying issue is that "verifier" may be ambiguous here. Is it the PCS verifier, or Plonk verifier? 
The verfier in plonk acts as PCS prover (computing PI commitments) and PCS verifier, so things get confusing...

iquerejeta replied:
I will implement that with #19 . Indeed, we only need [s]_2 for the verifier params of PCS. The positive aspect here is that we do not QUERY instances (we've fixed that to false), so the verifier does not need to commit to the public polynomials. This allows for a much simpler PCS verifier key.

davidnevadoc commented on src/poly/kzg/params.rs:
```diff
@@ -298,81 +294,49 @@ where
 
-    fn new(k: u32) -> Self {
+    /// UNSAFE function - do not use in production
+    pub fn new(k: u32) -> Self {
```
I'd either remove this function altogether, or rename it to something like `dummy_parameters` that clearly indicates that these are not usable parameters.

This used to be shared with parameters that could actually be generated, like IPA parameters. So it is here for "historic" reasons really.

iquerejeta replied:
We still need to be able to generate the parameters of the commitment scheme, otherwise we cannot do generic tests. I've moved that change to the other PR [here](https://github.com/input-output-hk/halo2/pull/21/files#diff-ee3c2408bfbf948ebc0612573d2488bdbdcb044405261f1594fde49c26f26485R22-R26). I'm not convinced of that, but I cannot think about a better solution.

davidnevadoc commented on src/poly/kzg/params.rs:
```diff
@@ -298,81 +294,49 @@ where
-    }
-
-    fn get_g(&self) -> &[E::G1Affine] {
+    /// TODO
```
Leftover?
A couple of TODOs across the file that I think are just leftovers.

davidnevadoc commented on src/plonk/permutation/verifier.rs:
```diff
@@ -192,63 +203,65 @@ impl<C: CurveAffine> Evaluated<C> {
-impl<C: CurveAffine> CommonEvaluated<C> {
-    pub(in crate::plonk) fn queries<'r, M: MSM<C> + 'r>(
+impl<F: PrimeField> CommonEvaluated<F> {
+    // todo: MSM trick here
```
todo!

davidnevadoc commented on src/poly/query.rs:
```diff
@@ -0,0 +1,146 @@
+
+/// A polynomial query at a point
+#[derive(Debug, Clone, Copy)]
+pub struct VerifierQuery<F: PrimeField, CS: PolynomialCommitmentScheme<F>> {
```
This is so much better than the previous version, nice!

miguel-ambrona replied:
So we need the field because it is a generic of the PCS, I see.

Alternatively, the PCS may take no generics and we use `PCS::Field` wherever we need `F`.

iquerejeta replied:
But if field is an associated type to PCS, then we need to have a different structure for each field

miguel-ambrona replied:
I think `<F, PCS<F>>` is very good, let's keep it.

davidnevadoc commented on src/poly/query.rs:
```diff
@@ -0,0 +1,146 @@
+    }
+}
+
+// #[allow(clippy::upper_case_acronyms)]
```
leftover

davidnevadoc commented on src/plonk/verifier.rs:
```diff
@@ -0,0 +1,308 @@
+use crate::poly::VerifierQuery;
+use crate::transcript::{read_n, Hashable, Sampleable, Transcript};
+
+/// Returns a boolean indicating whether or not the proof is valid
```
Returns Ok if the proof is valid, or an error otherwise ( not a boolean ).

This function has changed a lot. It previously received some kind of accumulator and updated it with a new proof, returning the same type of accumulator. I like this better, as it is way more intuitive than the previous monster. How do we handle accumulation now?

iquerejeta replied:
For now I've split that into a different PR that will implement #19

davidnevadoc replied:
I think it is good that we add the batch verification on a separate PR.
I just meant that the doc doesn't agree with the code because we don't return a boolean :sweat_smile:

iquerejeta replied:
ah, fair point. Handled in the next PR as well.

davidnevadoc reviewed (commented):
First pass

iquerejeta committed:
Remove issue template

iquerejeta committed:
Review comments

miguel-ambrona commented on src/poly/commitment.rs:
```diff
@@ -1,245 +1,40 @@
+    /// Parameters needed to verify a proof in the PCS
+    type VerifierParameters: From<Self::Parameters>;
+    /// Type of a committed polynomial
+    type Commitment: Clone + Copy + PartialEq + SerdeObject + Send + Sync; // Thinking of having an MSM here
```
Where is PartialEq needed?

miguel-ambrona replied:
Is Copy needed?

iquerejeta replied:
Copy is needed for simplicity (we can move out of a reference without cloning) and partialEq is needed for queries (we group evaluation points by their commitments, so we need that comparison)

miguel-ambrona replied:
I was asking about Copy because we may want to let the user be aware of the fact they are copying memory.

davidnevadoc replied:
I don't think I follow here :thinking: 
Are you concerned about other commitment schemes? I think this is completely fine for KZG.

miguel-ambrona commented on src/poly/commitment.rs:
```diff
@@ -1,245 +1,40 @@
+    /// Setup the parameters for the PCS
+    fn setup(k: u32) -> Self::Parameters;
+    /// Create a new instantiation of the PCS
+    fn new(params: Self::Parameters) -> Self;
```
Not sure what this is when I first see it.

miguel-ambrona replied:
I would leave it for now, but it could be better to pass the `Parameters` to `commit` and `open` instead of passing `self`. Then removing this `new` function.

iquerejeta replied:
I'll open an issue to address it in a different PR. https://github.com/input-output-hk/halo2/issues/20

miguel-ambrona commented on src/poly/kzg/mod.rs:
```diff
@@ -25,50 +25,47 @@ use halo2curves::serde::SerdeObject;
     type Commitment = E::G1Affine;
 
+    /// Unsafe function - do not use in production
     fn setup(k: u32) -> ParamsKZG<E> {
```
Shall this be under a flag for tests?

iquerejeta replied:
In a new PR I'm working I'm removing most of the setup functions, and I'm removing one for convenience called `unsafe_setup` or something along those lines.

miguel-ambrona commented on examples/shuffle.rs:
Would it make sense to remove this example?

miguel-ambrona commented on Cargo.toml:
```diff
@@ -8,13 +8,12 @@ authors = [
-rust-version = "1.66.0"
+rust-version = "1.76.0"
 description = """
-Fast PLONK-based zero-knowledge proving system with no trusted setup
```
Good catch! :D

miguel-ambrona commented on examples/proof-size.rs:
```diff
@@ -89,7 +89,7 @@ fn main() {
     let circuit = TestCircuit {};
 
     let model = from_circuit_to_model_circuit::<_, _, 56, 56>(
```
Wait, those 56 look sus.

iquerejeta replied:
Good catch

miguel-ambrona commented on src/dev.rs:
```diff
@@ -820,7 +823,15 @@ impl<F: FromUniformBytes<64> + Ord> MockProver<F> {
-                                                if r.cells.contains_key(&(cell.column, cell_row)) {
+                                                if r.cells.contains_key(&(cell.column, cell_row))
+                                                    || gate.polynomials().par_iter().all(|expr| {
+                                                        self.cell_is_irrelevant(
```
Love this.

miguel-ambrona commented on src/poly.rs:
```diff
@@ -4,33 +4,31 @@
 
-/// Generic commitment scheme structures
-pub mod commitment;
+// /// Generic commitment scheme structures
+// pub mod commitment;
```
Leftover?

miguel-ambrona commented on src/poly.rs:
```diff
@@ -4,33 +4,31 @@
-
-/// Inner product argument commitment scheme
-pub mod ipa;
+// mod strategy;
```
Remove?

miguel-ambrona commented on src/poly.rs:
```diff
@@ -4,33 +4,31 @@
 pub use domain::*;
 pub use query::{ProverQuery, VerifierQuery};
-pub use strategy::{Guard, VerificationStrategy};
+// pub use strategy::{Guard, VerificationStrategy};
```
Remove?

miguel-ambrona commented on tests/plonk_api.rs:
Leftover?

miguel-ambrona reviewed (approved):
This is great! Some would call it a piece of art.

Mind some leftovers of commented code or TODOs. I think David and I marked them all.

iquerejeta committed:
Bring back using the SerdeFormat for keys

iquerejeta committed:
Address review comments

iquerejeta committed:
Leftover docs
